feat: Support default_values for Select of type ComponentType.mentionable_select and ComponentType.channel_select

doluk opened the pull request:
## Summary
Looking at https://discord.com/developers/docs/components/reference#user-select I noticed that there is support for preselecting users, channels, roles, mentionables for the non string selects. This pr adds support for this.

## Information

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

doluk commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+        *,
+        default_value: User | Member | Role | GuildChannel | Thread,
+    ):
+        """Adds a default value to the select menu.
```
While implementing I considered modifying the SelectOption to make it work in this case too. But that would be a bigger pain to make that work, so I decided to implement it with its own classes. But those are heavily inspired by the SelectOptions and how they are implemented. And they have this add_option and append_option function as well, so I went for it as well

Paillat-dev replied:
Ok

Paillat-dev replied:
Ok makes sense.

doluk commented on discord/types/components.py:
```diff
@@ -36,6 +36,7 @@
 ComponentType = Literal[1, 2, 3, 4]
 ButtonStyle = Literal[1, 2, 3, 4, 5, 6]
 InputTextStyle = Literal[1, 2]
+SelectMenuDefaultValueType = Literal["channel", "role", "user"]
```
From my understanding this was for type checking and creates some types for enums, so I tried to copy the implementation of other enums😅

Paillat-dev replied:
I think you can just import it from enums.py

Paillat-dev replied:
Unless that causes import chain issues, in that case leave it like this

Paillat-dev commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+            default_value_type = SelectMenuDefaultValueType.Role
+        if default_value_type is None:
+            raise InvalidArgument(
+                "default values have to be of type User, Member, Role or GuildChannel"
```
```suggestion
                "Default values have to be of type User, Member, Role or GuildChannel"
```

Paillat-dev commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+            raise InvalidArgument(
+                "default_values can only be set on non string selects"
+            )
+        if not isinstance(value, list):
```
The specific checks I'm talking about are the `list` and `all(isinstance)` ones. While I feel like the other might be justified, it feels redundant to have type checks with also typed parameters. Yes, to some extent it is good to have better exceptions, but python isn't strictly typed anyways, as of itself, and it feels a little arbitrary to add them here.

JustaSqu1d commented on discord/components.py:
```diff
@@ -494,6 +526,53 @@ def to_dict(self) -> SelectOptionPayload:
+    Attributes
+    ----------
+    id: :class:`int`
+        The snowflake id of the default option.
```
```suggestion
        The snowflake ID of the default option.
```

JustaSqu1d commented on discord/enums.py:
```diff
@@ -1078,6 +1079,17 @@ class SubscriptionStatus(Enum):
     inactive = 2
 
 
+class SelectMenuDefaultValueType(Enum):
+    """The type of a select menu's default value."""
+
+    Channel = "channel"
+    Role = "role"
+    User = "user"
+
+    def __str__(self):
+        return self.name
+
+
```
Add enums to docs

JustaSqu1d commented on discord/ui/select.py:
```diff
@@ -134,13 +135,80 @@ def __init__(
+        if default_values:
+            if select_type is ComponentType.string_select:
+                raise InvalidArgument(
+                    "default_values parameter is only valid for user/role/channel/mentionable selects"
```
```suggestion
                    "default_values parameter is only valid for SelectMenu of type user, role, channel, or mentionable"
```

JustaSqu1d commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+        Raises
+        ------
+        ValueError
+                The number of options exceeds 25.
```
```suggestion
            The number of options exceeds 25.
```

JustaSqu1d commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+        Parameters
+        ----------
+        default_value: :class:`discord.SelectDefaultValue`
+                The default value to append to the select menu.
```
```suggestion
            The default value to append to the select menu.
```

JustaSqu1d commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+        Raises
+        ------
+        ValueError
+                The number of options exceeds 25.
```
```suggestion
            The number of options exceeds 25.
```

doluk commented:
I addressed most of the review comments, the two left open are something which need a final decision on what to do. @Paillat-dev @JustaSqu1d 
Hope the example I added is suitable

Paillat-dev commented on discord/ui/select.py:
```diff
@@ -324,6 +393,114 @@ def append_option(self, option: SelectOption):
+            raise InvalidArgument(
+                "default values have to be of type GuildChannel or Thread"
+            )
+        elif self.type is ComponentType.channel_select:
```
@doluk Anything about this ?

Paillat-dev commented:
@doluk I merged it but I am not sure it is all good, would you mind taking a look ?

DA-344 commented on discord/ui/select.py:
```diff
@@ -334,6 +404,118 @@ def append_option(self, option: SelectOption) -> Self:
+
+        Parameters
+        ----------
+        default_value: Union[:class:`discord.User`, :class:`discord.Member`, :class:`discord.Role`,
+        :class:`discord.abc.GuildChannel`, :class:`discord.Thread`]
+            The to be added default value
```
```suggestion
        default_value: Union[:class:`discord.User`, :class:`discord.Member`, :class:`discord.Role`,
        :class:`discord.abc.GuildChannel`, :class:`discord.Thread`, :class:`discord.Object`]
            The to be added default value. If this is an ``Object`` instance, you must pass ``type`` if the select type is ``mentionable``.
        type: :class:`discord.SelectMenuDefaultvalueType`
            The default value type, required if ``default_value`` is an ``Object`` and the select type is ``mentionable``.
```

DA-344 commented on discord/ui/select.py:
```diff
@@ -334,6 +404,118 @@ def append_option(self, option: SelectOption) -> Self:
+        elif self.type is ComponentType.channel_select:
+            default_value_type = SelectMenuDefaultValueType.Channel
+        if self.type is ComponentType.user_select and not isinstance(
+            default_value, (User, Member)
+        ):
+            raise InvalidArgument(
+                "Default values have to be of type User or Member with type ComponentType.user_select"
+            )
+        elif self.type is ComponentType.user_select:
+            default_value_type = SelectMenuDefaultValueType.User
+        if self.type is ComponentType.role_select and not isinstance(
+            default_value, Role
+        ):
+            raise InvalidArgument(
+                "Default values have to be of type Role with type ComponentType.role_select"
+            )
+        elif self.type is ComponentType.role_select:
+            default_value_type = SelectMenuDefaultValueType.Role
+        if self.type is ComponentType.mentionable_select and not isinstance(
+            default_value, (User, Member, Role)
+        ):
+            raise InvalidArgument(
+                "Default values have to be of type User, Member or Role with type ComponentType.mentionable_select"
+            )
+        elif self.type is ComponentType.mentionable_select and isinstance(
+            default_value, (User, Member)
+        ):
+            default_value_type = SelectMenuDefaultValueType.User
+        elif self.type is ComponentType.mentionable_select and isinstance(
+            default_value, Role
+        ):
+            default_value_type = SelectMenuDefaultValueType.Role
+        if default_value_type is None:
+            raise InvalidArgument(
+                "Default values have to be of type User, Member, Role or GuildChannel"
+            )
+        default_value = SelectDefaultValue(
+            id=default_value.id,
+            type=default_value_type,
+        )
```
a mix of what paillat said and support for passing Object instances.
```suggestion
        if self.type is ComponentType.mentionable_select and isinstance(default_value, Object) and type is MISSING:
            raise ValueError('type is required when the default_value is an Object and select type is mentionable')
        elif self.type is ComponentType.mentionable_select and isinstance(default_value, (Member, User)):
            default_value_type = SelectMenuDefaultValueType.user
        elif self.type is ComponentType.mentionable_select and isinstance(default_value, Role):
            default_value_type = SelectMenuDefaultValueType.role
        elif self.type is ComponentType.mentionable_select:
            default_value_type = type
        else:
            value_mapping = {
                ComponentType.role_select: SelectMenuDefaultValueType.role,
                ComponentType.channel_select: SelectMenuDefaultValueType.channel,
                ComponentType.user_select: SelectMenuDefaultValueType.user,
            }
            
            try:
                default_value_type = value_mapping[self.type]
            except KeyError:
                raise ValueError(f'invalid select type: {self.type}')

        default_value = SelectDefaultValue(
            id=default_value.id,
            type=default_value_type,
        )
```

DA-344 commented on discord/ui/select.py:
```diff
@@ -30,9 +30,9 @@
 from ..channel import _threaded_guild_channel_factory
-from ..components import SelectMenu, SelectOption
+from ..components import SelectDefaultValue, SelectMenu, SelectOption
 from ..emoji import AppEmoji, GuildEmoji
-from ..enums import ChannelType, ComponentType
+from ..enums import ChannelType, ComponentType, SelectMenuDefaultValueType
 from ..errors import InvalidArgument
 from ..interactions import Interaction
 from ..member import Member
```
The changes i've suggested require importing Object 
```suggestion
from ..components import SelectDefaultValue, SelectMenu, SelectOption
from ..emoji import AppEmoji, GuildEmoji
from ..enums import ChannelType, ComponentType, SelectMenuDefaultValueType
from ..errors import InvalidArgument
from ..interactions import Interaction
from ..member import Member
from ..object import Object
```
